### PR TITLE
[bgp] Add design doc for explicitly configuring BGP router id

### PR DESCRIPTION
**Pull request by yaqiangz**
This document describes the design details of explicitly configuring BGP router id in SONiC.

| Description | Repo | PR title | State |
| ------ | ------ | ------- | ----- |
| Update yang model | sonic-buildimage |https://github.com/sonic-net/sonic-buildimage/pull/18534| GitHub issue/pull request detail |
| Add support for minigraph parser | sonic-buildimage |https://github.com/sonic-net/sonic-buildimage/pull/18698| GitHub issue/pull request detail |
| Update bgpcfgd and configuration j2 for single-asic | sonic-buildimage | https://github.com/sonic-net/sonic-buildimage/pull/18727 |GitHub issue/pull request detail|
| Update bgpcfgd and configuration j2 for multi-asic | sonic-buildimage ||To be done|

### CODE REVIEW

File: doc/BGP/BGP-router-id.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+This document describes a mechanism that to use customized router id when Loopback IPv4 address is not configured.
+
+### Overview
+
+Currently, there is hard coding in SONiC to use Loopback IPv4 address to be BGP router id [bgp.main.conf.j2](https://github.com/sonic-net/sonic-buildimage/blob/master/dockers/docker-fpm-frr/frr/bgpd/bgpd.main.conf.j2#L91), which would cause BGP cannot work when IPv4 address in Loopback doesn't configure.
+
+### Requirements
+
+Add support to customize BGP router id.
```
Comment (Blueve) [inline doc/BGP/BGP-router-id.md:35]: Add support to allow user explicitly configure BGP router id.

File: doc/BGP/BGP-router-id.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+### High Level Design
+
+1. Add field `bgp_router_id` into `CONFIG_DB["DEVICE_METADATA"]["localhost"]["bgp_router_id"]`
+2. In [bgp.main.conf.j2](https://github.com/sonic-net/sonic-buildimage/blob/master/dockers/docker-fpm-frr/frr/bgpd/bgpd.main.conf.j2#L91) to always respect bgp_router_id in CONFIG_DB if configured
```
Comment (Blueve) [inline doc/BGP/BGP-router-id.md:40]: I don't think we need discuss detail implementation here.
Form high-level design, let's discuss different situation here:
||Loopback0 IPv4 address exists|Loopback0 IPv4 address not exists|
|---|---|---|
|bgp_router_id configured|honor bgp_router_id|honor bgp_router_id|
|bgp_router_id not configured|honor Loopback0 IPv4|FRR default router ID value is selected as the largest IP Address of the interfaces. When router zebra is not enabled _bgpd_ can’t get interface information so router-id is set to 0.0.0.0|

File: doc/BGP/BGP-router-id.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+### Definitions/Abbreviations
+
+| Definitions/Abbreviation | Description |
+| ------------------------ | ----------- |
+| FRR | A free and open source Internet routing protocol suite for Linux and Unix platforms |
+| Router ID | 32-bit value that uniquely identifies a BGP device |
+
+### Scope
+
+This document describes a mechanism that to use customized router id when Loopback IPv4 address is not configured.
```
Comment (Blueve) [inline doc/BGP/BGP-router-id.md:27]: This statement is not aligned with #2 item in `High Level Design`.

This design is to allow user to explicitly configure BGP router id.

File: doc/BGP/BGP-router-id.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+### Overview
+
+Currently, there is hard coding in SONiC to use Loopback IPv4 address to be BGP router id [bgp.main.conf.j2](https://github.com/sonic-net/sonic-buildimage/blob/master/dockers/docker-fpm-frr/frr/bgpd/bgpd.main.conf.j2#L91), which would cause BGP cannot work when IPv4 address in Loopback doesn't configure.
```
Comment (Blueve) [inline doc/BGP/BGP-router-id.md:31]: BGP router id was defined as a 32-bit value that uniquely identifies a BGP device. Currently, SONiC uses **Loopback0** IPv4 address as BGP router id. This coupling prevents users from using customized router id.

Comment (Blueve) [inline doc/BGP/BGP-router-id.md:31]: The link is not a permlink

Comment (Blueve) [inline doc/BGP/BGP-router-id.md:31]: Have you confirmed BGP cannot work if no Loopback0 IPv4 address configured on device? I thought BGP can still work if FRR selected router-id not conflicting with whole BGP network.

File: doc/BGP/BGP-router-id.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+### Requirements
+
+Add support to allow user explicitly configure BGP router id.
+
+### High Level Design
+
+2 aspects enhancement:
+
+1. Add a field `bgp_router_id` in `CONFIG_DB["DEVICE_METADATA"]["localhost"]` to support explicitly configure BGP router id. If `CONFIG_DB["DEVICE_METADATA"]["localhost"]["bgp_router_id"]` configured, always use it as BGP router id no matter whether Lo IPv4 exist.
```
Comment (Blueve) [inline doc/BGP/BGP-router-id.md:55]: Lo IPv4 -> Loopback0 IPv4

File: doc/BGP/BGP-router-id.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+| -------- | ----------- | ---------------- | ------------------ |
+| 1.0      | Mar 27 2024 | Yaqiang Zhu, Jing Kan | Initial proposal   |
+
+### Definitions/Abbreviations
+
+| Definitions/Abbreviation | Description |
+| ------------------------ | ----------- |
+| FRR | A free and open source Internet routing protocol suite for Linux and Unix platforms |
+| BGP Router ID | 32-bit value that uniquely identifies a BGP device |
+| Lo | Loopback interface in SONiC |
```
Comment (Blueve) [inline doc/BGP/BGP-router-id.md:24]: This is unnecessary, the change is specific to Loopback0

**Review by Blueve — APPROVED** [review]

File: doc/BGP/BGP-router-id.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+bgp_router_id   = inet:ipv4-address         ; Customized BGP router id
+```
+
+**Sample of CONFIG DB snippet:**
+
+```json
+{
+    "DEVICE_METADATA": {
+        "localhost": {
+            "bgp_router_id": "10.1.0.32"
```
Comment (prsunny) [inline doc/BGP/BGP-router-id.md:103]: Is this going to be same as loopback IP?

Comment (yaqiangz) [reply doc/BGP/BGP-router-id.md:103]: Not mandatory, this configuration depends on the user. The user need to make sure that the bgp router id configured is unique in the AS

File: doc/BGP/BGP-router-id.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+|          | Loopback0 IPv4 address exists | Loopback0 IPv4 address doesn't exist |
+|--------------|-------|------------|
+| bgp_router_id configured | Add BGP peer | Add BGP peer |
+| bgp_router_id doesn't be configured | Add BGP peer | Do not add BGP peer |
+
+Below is new workflow, the main changes are in `1.` and `3.`.
+
+1. After bgp container started, configuration file `/etc/frr/bgpd.conf` for bgpd is would be rendered.
+   * If CONFIG_DB`["DEVICE_METADATA"]["localhost"]["bgp_router_id"]` exists, use it as BGP router id.
```
Comment (qiluo-msft) [inline doc/BGP/BGP-router-id.md:71]: Will it work multi-asic?

File: doc/BGP/BGP-router-id.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+   * Else if Loopback0 IPv4 address exists, use it as BGP router id.
+   * Else, BGP router id wouldn't be specified.
```
Comment (qiluo-msft) [inline doc/BGP/BGP-router-id.md:73]: Should we break and report failure in the final "else" branch? Should we allow device working in this way?

Comment (qiluo-msft) [reply doc/BGP/BGP-router-id.md:73]: If we consider this case as critical failure, we may need to add constrain in yang model.

File: doc/BGP/BGP-router-id.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+### Overview
+
+Currently, there are some BGP hard codings in SONiC:
+1. BGP router id was defined as a 32-bit value that uniquely identifies a BGP device. SONiC uses Loopback0 IPv4 address as BGP router id. This coupling prevents users from using customized router id. And FRR would choose an IP address in device to be BGP router id if Loopback0 IPv4 address doesn't exist. If the router id choosen by FRR is not unique in AS, it would be considered an error.
+2. SONiC wouldn't add BGP peer when there is not Loopback0 IPv4 exists. It would cause BGP cannot establish.
+
+Below is current workflow about BGP and router id, only includes contents related to Loopback0.
+
+1. After bgp container started, configuration file `/etc/frr/bgpd.conf` for bgpd would be rendered. It will use Loopback0 IPv4 address as BGP router id, if it doesn't exist, the BGP router id wouldn't be specified.
+2. bgpd start with configuration rendered before. If BGP router id is not specified, it would choose an IP address in device to be BGP router id.
```
[the remark this answers is not in the document]
Comment (Blueve) [reply doc/BGP/BGP-router-id.md:39]: The highest IP will be choosed. This is FRR's behaviour.

File: doc/BGP/BGP-router-id.md (hunk 9 above)
Comment (Blueve) [reply doc/BGP/BGP-router-id.md:73]: This is existing behaviour. The design is trying to not make breaking change.

Existing behaviour:
bgpcfgd will not adding neighbors if router id not specified and bgp is actually not working.

File: doc/BGP/BGP-router-id.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+### Requirements
+
+Add support to allow user explicitly configure BGP router id.
+
+### High Level Design
+
+2 aspects enhancement:
+
+1. Add a field `bgp_router_id` in `CONFIG_DB["DEVICE_METADATA"]["localhost"]` to support explicitly configure BGP router id. If `CONFIG_DB["DEVICE_METADATA"]["localhost"]["bgp_router_id"]` configured, always use it as BGP router id no matter whether Loopback0 IPv4 exist. With this change, the new BGP router id configuration behavior will be like follow. To be clarified that when bgp_router_id doesn't be configured, the behavior is totally same as previously.
```
Comment (xincunli-sonic) [inline doc/BGP/BGP-router-id.md:54]: In a multi-ASIC environment, would each ASIC have its own `bgp_router_id`?

File: doc/BGP/BGP-router-id.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+2. bgpd start with configuration rendered before. If BGP router id is not specified, it would choose an IP address in device to be BGP router id.
+3. After bgpcfgd started, it will add bgp peer depends on whether Loopback0 IPv4 exist.
+
+<p align=center>
+<img src="img/origin_bgp_seq.png" alt="Figure 1. Origin bgp seq" width=700>
+</p>
+
+### Requirements
+
+Add support to allow user explicitly configure BGP router id.
```
Comment (xincunli-sonic) [inline doc/BGP/BGP-router-id.md:48]: In multi-ASIC architectures, inter-ASIC BGP sessions often rely on specific Loopback interfaces or unique IP addresses assigned per ASIC for peering. Could we explicitly address how these scenarios are managed, especially considering that BGP sessions within the device (for the purpose of route exchange between ASICs) and external BGP sessions might have different requirements for identifying and using Router IDs.

File: doc/BGP/BGP-router-id.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+| bgp_router_id doesn't be configured | Honor Loopback0 IPv4 address | FRR default router ID value is selected as a IP Address of the device. When router zebra is not enabled bgpd can’t get interface information so router-id is set to 0.0.0.0 |
+
+2. Remove strong dependencies on Loopback0 IPv4 address when adding BGP peer in the situation that bgp_router_id is configured.With this change, the new BGP peer adding behavior will be like follow. To be clarified that when bgp_router_id doesn't be configured, the behavior is totally same as previously.
+
+|          | Loopback0 IPv4 address exists | Loopback0 IPv4 address doesn't exist |
+|--------------|-------|------------|
+| bgp_router_id configured | Add BGP peer | Add BGP peer |
+| bgp_router_id doesn't be configured | Add BGP peer | Do not add BGP peer |
+
+Below is new workflow, the main changes are in `1.` and `3.`.
```
Comment (xincunli-sonic) [inline doc/BGP/BGP-router-id.md:68]: The design doesn't provide a clear mechanism for managing this in the context of multiple ASICs, where configurations might need to be more granular or ASIC-specific.

File: doc/BGP/BGP-router-id.md (hunk 8 above)
Comment (yaqiangz) [reply doc/BGP/BGP-router-id.md:90]: Yes, improved the multi-asic part of HLD

File: doc/BGP/BGP-router-id.md (hunk 11 above)
Comment (yaqiangz) [reply doc/BGP/BGP-router-id.md]: Yes

File: doc/BGP/BGP-router-id.md (hunk 12 above)
Comment (yaqiangz) [reply doc/BGP/BGP-router-id.md:63]: Updated, currently FrontEnd asic use Loopback4096 as router-id for iBGP and eBGP. BackEnd asic use Loopback4096 as router-id for iBGP. Hence we can specify router-id in config_db of correspond asic

File: doc/BGP/BGP-router-id.md (hunk 13 above)
Comment (yaqiangz) [reply doc/BGP/BGP-router-id.md]: Updated

**Review by Blueve — APPROVED** [review]

File: doc/BGP/BGP-router-id.md (hunk 9 above)
Comment (qiluo-msft) [reply doc/BGP/BGP-router-id.md:93]: Existing behaviour does not always mean right behavior. If we consider this case as critical failure, we may need to add constrain in yang model.

File: doc/BGP/BGP-router-id.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply doc/BGP/BGP-router-id.md:41]: FRR's behaviour seems a random algorithm to me. Is it working in production?

[the remark this answers is not in the document]
Comment (Blueve) [reply doc/BGP/BGP-router-id.md:41]: No. We always specify Loopback0 and never fall back to default FRR behavior.
And the BGP neighbors will not even been added if router id not specified.

**Review by Blueve — APPROVED** [review]

**Review by Blueve — APPROVED** [review]

**Review by Blueve — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**